pkg/tlscert/self_signed.go

```diff
@@ -0,0 +1,99 @@
+package tlscert
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"time"
+)
+
+// NewSelfSignedCertProvider creates a new certificate provider that creates a self-signed certificate.
+func NewSelfSignedCertProvider(commonName string, dnsNames []string, orgs []string) SelfSignedProvider {
+	return SelfSignedProvider{commonName: commonName, dnsNames: dnsNames, orgs: orgs}
+}
+
+type SelfSignedProvider struct {
+	orgs       []string
+	dnsNames   []string
+	commonName string
+}
+
+func (p SelfSignedProvider) Certificate(context.Context) (Certificate, error) {
+	// init CA config
+	ca := &x509.Certificate{
+		SerialNumber:          big.NewInt(2022),
+		Subject:               pkix.Name{Organization: p.orgs},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(1, 0, 0), // expired in 1 year
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	// generate private key for CA
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return Certificate{}, err
+	}
+
+	// create the CA certificate
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivateKey.PublicKey, caPrivateKey)
+	if err != nil {
+		return Certificate{}, err
+	}
+
+	// CA certificate with PEM encoded
+	caPEM := new(bytes.Buffer)
+	_ = pem.Encode(caPEM, &pem.Block{
```

Comment:
Can you add a comment explaining why this would never error? (or can we just handle the error?)

some comment appies in two places further down

Reply:
This error is the error of writing into the bytes.Buffer, so it basically can never happen. Made that explicit in https://github.com/grafana/rollout-operator/pull/25/commits/4dc632c6e19b32bb1a55afac6f4e773c3c8024ae by panicking with proper message.